### ISSUE TITLE
cmake extended_apis not found

### ISSUE THREAD

**Issue by Outstep** [Bareflank/hypervisor#872]
Hello,

I was trying to build the EFI version using the video:

https://www.youtube.com/watch?v=FuEyjDqA53M&t=4s

but the 

hypervisor/scripts/cmake/config/example_config.cmake

is different and there is not setting to "ENABLE_EXTENDED_APIS=ON" found.

Has something changed since the video was made?
Thanks

**Comment by JaredWright** [comment]
Yes, quite a bit has changed in the project's build system since that video was created. The "extended apis" used to be a Bareflank extension, but has since been deprecated and merged into the base "hypervisor" repository. 

I verified just now that I could build build Bareflank as an EFI application using the configuration file ```hypervisor/scripts/cmake/config/example_config.cmake```, except I changed line 66 from:

```
set(ENABLE_BUILD_EFI OFF)
```

to this instead:

```
set(ENABLE_BUILD_EFI ON)
```

Does that help you get started?

**Comment by Outstep** [comment]
Hi Jared,

Thanks for getting back to me on this.

Do I still need to change the config.cmake:

# Override VMM
set(OVERRIDE_VMM <name>)

# Override VMM Target
set(OVERRIDE_VMM_TARGET <name>)

If yes, then I cannot locate the

bfvmm/integration/arch/intel_x64/efi/CMakeLists.txt which does not seem to
exist, so I cannot compile the integration tests either.

Also, you have some dead links on the main github page in the paragraph:

"
For more detailed build instructions, see the detailed build instructions
<https://github.com/Bareflank/hypervisor/blob/master/scripts/docs/build_instructions.md>.
For instructions on building and creating Bareflank extensions, see
the extension
build instructions
<https://github.com/Bareflank/hypervisor/blob/master/scripts/docs/extension_instructions.md>
"

Cheers,
Lonnie

**Comment by JaredWright** [comment]
Thanks for the tip on the broken links! There is a huge change set we're anticipating coming into the project this month, and we will likely have to address many issues in our instructions after that change set gets merged. Thanks for bearing with us during the turbulence...

The build configurations:

```
# Override VMM
set(OVERRIDE_VMM <name>)

# Override VMM Target
set(OVERRIDE_VMM_TARGET <name>)
```

control which VMM the build system will turn into an EFI application, because the build system is capable of compiling many VMMs at once. Are you getting build errors when you try to set those configurations to your own custom extension?

**Comment by Outstep** [comment]
Hi Jared,

Hope that your day is starting out well.

Yes, I learned about the configuration variables in the YouTube video that
I watch.  My problem now, I that I do not know what to set them to for my
Ubuntu 18.04 (64-bit) system and am not sure what you mean when you said
"your own custom extension" as I am just trying to replicate what the video
was doing in building an EFI image for Linux & Windows and then booting it
and into one of the OS's after which you could make a Hypercall to
BareFlank from within Linux, for example to get back some information.

Before I start developing on it, I wanted to be able to basically replicate
that you had done as a a demo.
Cheers,
Lonnie

**Comment by Outstep** [comment]
Hi Jared,

Hope that all is well.

I am still having problems building things on BareFlank and just remembered
that you mentioned earlier that there would be a lot of new updates and
fixes this month.  I was wondering if you had a projection as to when it
all might be ready since I think that I should just wait until the new
version is ready so that I can follow some examples to get a feel for how
to build BareFlank. In particular, I am greatly interested in the EFI
version since that seems to be the closest starting point towards a type of
Separation Kernel approach in that, from what I saw in the video, it would
allow the hypervisor to be booted (EFI level) and then chain load booting
into either Windows ( guessing Windows 7 Pro, Windows 10, etc.. and
hopefully in 64-bit) or chain booting Linux (I have Ubuntu 18.04 64-bit).

My thoughts are that if I can get the the basic EFI demo working and chain
booting one of these basic OS's then I can start to explore how to possibly
have the hypervisor report and partition off just a sub-set of the
resources (CPU's, RAM, etc...) such that when it chain boots an OS then it
would only have part of the resources. Then, I can explore about
implementing the hypervisor level context switching to allow for multiple
OS's running on native hardware without a heavy VMM. The VMM can be
investigated later when I want to virtualize to many OS's on all or part of
the total system resources in the virtualization phase, but that is later.

Thanks again for all of your help and for taking the time to answer my
newbie questions in this arena.
Lonnie

**Comment by JaredWright** [comment]
@rianquinn Any sense of an ETA on when your mega-PR will be coming in?

@Outstep If the changes won't be ready for more than a couple weeks, I'll dedicate some time this weekend to helping you get started with building Bareflank in an EFI configuration.

**Comment by rianquinn** [comment]
I will test master tonight on creating an extension and running UEFI with it. Everything should be working, but I want to verify that it is before commenting further with updated instructions. 

Once the mega-PR is done, I will re-create the videos so that they are up-to-date.

**Comment by rianquinn** [comment]
Ok, everything worked fine for me. 

Here is my bash history
```bash
  968  mkdir test
  969  cd test/
  970  git clone https://github.com/Bareflank/hypervisor.git
  971  mkdir build
  972  mkdir cache
  973  cp hypervisor/scripts/cmake/config/example_config.cmake config.cmake
  974  vim config.cmake 
  975  cd build/
  976  cmake ../hypervisor/
  977  make -j5
```

And here is my config.cmake

```
...
# ------------------------------------------------------------------------------
# Options
# ------------------------------------------------------------------------------

# Boxy
#
# This option enables the use of the boxy. It assumes that
# boxy is located in the same directory as this configuration file as stated
# in the example working directory above.
#
set(ENABLE_BOXY OFF)

# Enable EFI
#
# This will enable building EFI targets after the VMM has compiled. Note that
# this only works on Linux (or the Linux subsystem for Windows)
#
set(ENABLE_BUILD_EFI ON)

# Developer Mode
#
# Turns on build options useful for developers. If you plan to submit a PR to
# any of the Bareflank repos, this option will be needed as it enables
# formatting, static / dynamic analysis, etc...
#
set(ENABLE_DEVELOPER_MODE OFF)

# ------------------------------------------------------------------------------
# Override Options
# ------------------------------------------------------------------------------

# Override VMM
#
# Setting the OVERRIDE_VMM variable is the same as setting the DEFAULT_VMM
# from the command line. Use this to change which VMM the build system will
# load when you run 'make load` or `make quick`.
#
set(OVERRIDE_VMM test_efi)

# Override VMM Target
#
# This is only needed if you also turn on EFI. If you are building EFI from
# your own extension, you will need to tell the build system what the target
# name is for your VMM so that it will know what target the EFI portion of
# the build system depends on. This simply ensures that the build system
# first compiles your custom VMM before compiling the bareflank.efi. In other
# words, within the build system, this will add a call to add_dependency()
# with the value you provide.
#
set(OVERRIDE_VMM_TARGET bfvmm)

# Override Linux
#
# This is only useful when Boxy is enabled. This allows you to override the
# Linux repo that you are using. By default, if you do not set this variable,
# Boxy will download our version of Linux for you. If you are however a
# developer working on our version of Linux, you will need your own forked
# version of our Linux. This allows you to specify the location of your forked
# version so that Boxy will use your version instead of its default version.
#
# set(LINUX_DIR <path>)
...
```

This will produce prefixes/x86_64-efi-pe/bin/bareflank.efi. Put that file on your EFI partition and run it from an EFI shell. 

As for the massive patch set that is cominig, we are working hard to solidify our APIs now that we know what is needed for guest support, so ideally once this work is complete, our APIs will not change moving forward.

**Comment by rianquinn** [comment]
Oh, and one final note here. The "test_efi" is an integration test (one of many). All of the integration tests are "extensions". Feel free to look at the test to see what is needed in your extension to get EFI working (basically you have to enable EPT). To create your own extensions, please see our examples.

**Comment by Outstep** [comment]
Hi Rian,

Thanks for the steps. I followed your lead on this and was able to compile
things. Think that I was having a bit of confusion with what names to use
in the config.cmake.

I was able to compile the BareFlank and have the resulting efi files in:

~/test/build/bfvmm/x86_64-vmm-elf/build

I know that these are integration tests as you mentioned as well, but,
being just curious, what is the difference between:

test_efi
and
test_efi_with_interrupts

I will have to keep digging in more to learn how to do some things with
BearFlank but this will be a good start.

Thanks so very much.
Lonnie

**Comment by Outstep** [comment]
Just a follow up to on the results.

I placed the "test_efi" into my Ubuntu 18.04 (x86_64 actually AMD
processors with virtualization enabled)  /boot/efi directory and then
rebooted the system. Went to the BIOS utility shell and navigated to the
test_efi via the shell but the result was that it said that it could not
run test_efi.

Is there some additional step that I might have missed?
Thanks,
Lonnie

**Comment by rianquinn** [comment]
The test_efi is the VMM itself which is an ELF file. bareflank.efi is the UEFI application that has test_efi in it, plus the loader. You need to copy over bareflank.efi and run that. Not the test_efi.

**Comment by Outstep** [comment]
I could not locate any bareflank.efi file being generated in down my
"build" directory.  Where is it supposed to be located?
thanks

**Comment by Outstep** [comment]
May have just found it down the path:

build/prefixes/x86_64-efi-pe/bin

**Comment by rianquinn** [comment]
that's the location

**Comment by Outstep** [comment]
Hi Rian,

Sorry for the delay in responding.

Just tried to boot using the bareflank.efi and the resulting message was
that it completed with an error.

When I boot my system, I have an option to "run and efi program" but so far
have not seen a way to have a "efi shell" so I really could not gather any
information on the error code.

Any ideas on a way for me to debug this?

Do you know of any efi-shells that I might be able to try if there are any
such available?
Thanks,
Lonnie

**Comment by rianquinn** [comment]
You need to boot into an EFI shell. How to do that depends on your computer. On my ASUS motherboard, I have to download the EFI shell from Tiano's GitHub page and put the 64bit binary on a FAT formatted USB drive. From there, I can run bareflank.efi, all of which I demonstrate in the video. You will have to sort out how to do the same thing on your own system.

**Comment by Outstep** [comment]
Thanks. I'll look into that more to see what I can do on this end.

**Comment by JaredWright** [comment]
> I placed the "test_efi" into my Ubuntu 18.04 (x86_64 actually AMD
processors with virtualization enabled)

@Outstep Are you trying to run this on an AMD machine? At the moment, Bareflank supports Intel only.

**Comment by Outstep** [comment]
Yea, I actually think that is the problem. Perhaps I glossed over that
point if it was on your site, but right now all i have available is a an
AMD machine, but am thinking about getting an Intel soon.

**Comment by JaredWright** [comment]
Ah, bummer. AMD support is something that has been on the project's radar for a long time, but some other priorities are taking precedence at the moment.

Closing this issue. If you would like to discuss the project more, please feel free to open another issue or chat with us on Gitter!
